File: scala/src/main/scala/PrefixTrie.scala
import scala.collection.mutable

/**
 * Implementation A: Hash map trie (flexible alphabet).
 *
 * Each node stores children in a mutable.HashMap[Char, Node]. This
 * supports any character set without wasting memory on unused slots.
 * Lookup and insertion are O(L) where L is the word length, with O(1)
 * amortized per character via the hash map.
 */
class HashMapTrie {
  private val children = mutable.HashMap[Char, HashMapTrie]()
  private var isEnd = false

  def insert(word: String): Unit = {
    var current = this
    for (c <- word) {
      current = current.children.getOrElseUpdate(c, new HashMapTrie)
    }
    current.isEnd = true
  }

  def search(word: String): Boolean = {
    val node = traverse(word)
    node != null && node.isEnd
  }

  def startsWith(prefix: String): Boolean = {
    traverse(prefix) != null
  }

  private def traverse(s: String): HashMapTrie = {
    var current = this
    var i = 0
    while (i < s.length) {
      current.children.get(s(i)) match {
        case Some(child) => current = child
        case None => return null
      }
      i += 1
    }
    current
  }
}

/**
 * Implementation B: Fixed-array trie (lowercase ASCII only).
 *
 * Each node stores children in a fixed Array of 26 slots, one per
 * letter a-z. Lookup is a direct index: children(c - 'a'). This
 * trades memory (26 references per node regardless of usage) for
 * speed (no hashing, direct index access).
 *
 * For sparse tries with few children per node, the hash map version
 * uses less memory. For dense tries (e.g. dictionaries), the array
 * version is faster due to direct indexing.
 */
class ArrayTrie {
  private val children = new Array[ArrayTrie](26)
  private var isEnd = false

  def insert(word: String): Unit = {
    var current = this
    for (c <- word) {
      val idx = c - 'a'
      if (current.children(idx) == null) {
        current.children(idx) = new ArrayTrie
      }
      current = current.children(idx)
    }
    current.isEnd = true
  }

  def search(word: String): Boolean = {
    val node = traverse(word)
    node != null && node.isEnd
  }

  def startsWith(prefix: String): Boolean = {
    traverse(prefix) != null
  }

  private def traverse(s: String): ArrayTrie = {
    var current = this
    var i = 0
    while (i < s.length) {
      val idx = s(i) - 'a'
      if (current.children(idx) == null) return null
      current = current.children(idx)
      i += 1
    }
    current
  }
}

File: rust/src/prefix_trie.rs
use std::collections::HashMap;

/// Implementation A: Hash map trie (flexible alphabet).
///
/// Each node stores children in a HashMap<u8, Box<Node>>. Box provides
/// heap allocation for the recursive structure — without it, Node would
/// have infinite size. This supports any byte value as a character.
///
/// Rust's ownership model makes this straightforward: each node owns its
/// children exclusively via Box, so the entire trie is freed when the
/// root is dropped. No Rc<RefCell<>> needed because there is no shared
/// ownership — each child has exactly one parent.
pub struct HashMapTrie {
    children: HashMap<u8, Box<HashMapTrie>>,
    is_end: bool,
}

impl HashMapTrie {
    pub fn new() -> Self {
        Self {
            children: HashMap::new(),
            is_end: false,
        }
    }

    pub fn insert(&mut self, word: &str) {
        let mut current = self;
        for b in word.bytes() {
            current = current.children
                .entry(b)
                .or_insert_with(|| Box::new(HashMapTrie::new()));
        }
        current.is_end = true;
    }

    pub fn search(&self, word: &str) -> bool {
        match self.traverse(word) {
            Some(node) => node.is_end,
            None => false,
        }
    }

    pub fn starts_with(&self, prefix: &str) -> bool {
        self.traverse(prefix).is_some()
    }

    fn traverse(&self, s: &str) -> Option<&HashMapTrie> {
        let mut current = self;
        for b in s.bytes() {
            match current.children.get(&b) {
                Some(child) => current = child,
                None => return None,
            }
        }
        Some(current)
    }
}

/// Implementation B: Index-based trie with fixed array (lowercase ASCII only).
///
/// Nodes live in a Vec (pool), referenced by index rather than pointer.
/// Each node has a fixed array of 26 child indices (one per letter a-z),
/// where usize::MAX means "no child". This is the same index-based
/// pattern used in the LRU cache's manual linked list — it sidesteps
/// Rust's ownership constraints entirely since all nodes are owned by
/// the single Vec.
///
/// New nodes are allocated by pushing onto the pool. No free list is
/// needed because trie nodes are never removed in this implementation.
pub struct ArrayTrie {
    pool: Vec<ArrayTrieNode>,
}

struct ArrayTrieNode {
    children: [usize; 26],
    is_end: bool,
}

impl ArrayTrieNode {
    fn new() -> Self {
        Self {
            children: [usize::MAX; 26],
            is_end: false,
        }
    }
}

impl ArrayTrie {
    pub fn new() -> Self {
        Self {
            pool: vec![ArrayTrieNode::new()], // index 0 = root
        }
    }

    pub fn insert(&mut self, word: &str) {
        let mut current = 0;
        for b in word.bytes() {
            let idx = (b - b'a') as usize;
            if self.pool[current].children[idx] == usize::MAX {
                let new_idx = self.pool.len();
                self.pool.push(ArrayTrieNode::new());
                self.pool[current].children[idx] = new_idx;
            }
            current = self.pool[current].children[idx];
        }
        self.pool[current].is_end = true;
    }

    pub fn search(&self, word: &str) -> bool {
        match self.traverse(word) {
            Some(idx) => self.pool[idx].is_end,
            None => false,
        }
    }

    pub fn starts_with(&self, prefix: &str) -> bool {
        self.traverse(prefix).is_some()
    }

    fn traverse(&self, s: &str) -> Option<usize> {
        let mut current = 0;
        for b in s.bytes() {
            let idx = (b - b'a') as usize;
            if self.pool[current].children[idx] == usize::MAX {
                return None;
            }
            current = self.pool[current].children[idx];
        }
        Some(current)
    }
}
